tests/Integration/Behaviour/Features/Scenario/Product/category_tree.feature

```diff
@@ -0,0 +1,43 @@
+# ./vendor/bin/behat -c tests/Integration/Behaviour/behat.yml -s product --tags category-tree
+@reset-database-before-feature
+@clear-cache-before-feature
+@category-tree
+Feature: Show category tree in product page (BO)
+  As an employee
+  I need to be able to see category tree in product page with marked product-category associations
+
+  Background:
+    Given language with iso code "en" is the default one
+    And category "home" in default language named "Home" exists
+    And category "clothes" in default language named "Clothes" exists
+    And category "clothes" parent is category "home"
+    And category "men" in default language named "Men" exists
+    And category "men" parent is category "clothes"
+    And category "women" in default language named "Women" exists
+    And category "women" parent is category "clothes"
+    And category "accessories" in default language named "Accessories" exists
+    And category "accessories" parent is category "home"
+    And category "stationery" in default language named "Stationery" exists
+    And category "stationery" parent is category "accessories"
+    And category "home_accessories" in default language named "Home Accessories" exists
+    And category "home_accessories" parent is category "accessories"
+    And category "art" in default language named "Art" exists
+    And category "art" parent is category "home"
+
+  Scenario: I can see categories tree with product associations
+    Given I add product "product1" with following information:
+      | name[en-US] | unisex sneakers |
+      | type        | standard        |
+    And product "product1" type should be standard
+    And product "product1" should be assigned to default category
+    Then I should see following categories tree for product "product1" in "en" language:
+#     @todo: how to assert parent-child category tree?
```

Comment:
@matks any ideas how do I assert the tree ? :smile:

Reply:
Couldn't you check (direct) children for each parent ?

Reply:
I agree that checking the full tree will be a PITA especially how to find the correct way to represent it with behat syntax ^^ Besides you don't need the tree for Product associations, only a list of all associated categories

As I explained in my previous comments, you would need to different queries:
- `GetCategoriesTree` which returns the tree of categories, this one needs to represent the full tree
- `GetProductCategories` which returns the list of associated categories in a simple array, no need to reflect the tree in this result

This will make testing the product query much simpler, since it's a simple array

Reply:
To test the `GetCategoriesTree` the simpler will be to use different steps, let's say you have this kind of tree:

```
Root
  - Main Category 1
    - Sub category 1
    - Sub category 2
  - Main Category 2
    - Sub category 3
    - Sub category 4
```

The you would test:

```
Then the category root has these direct children
  Main category 1
  Main category 2
And the category Main category 1 has these direct children
  Sub category 1
  Sub category 2
And the category Main category 3 has these direct children
  Sub category 3
  Sub category 4
```

Reply:
For the tree you have two possibilities:
- run the query for each step and find the tested category, then assert its children It's simple but it requires running the query at each steps
- perform a step in two times as we did recently The first step runs the query `When I get category tree` and stores the result, and the second type of steps perform assertions `Then the category root has these direct children` and always uses the same result object